client/engine/engine.go

```diff
@@ -245,31 +265,102 @@ func (e *Engine) constructObjectiveFromMessage(message protocols.Message) (proto
+			return virtualfund.VirtualFundObjective{}, errors.New("a single hop virtual channel must have exactly 3 participants")
+		}
+
+		alice := participants[0]
+		intermediary := participants[1]
+		bob := participants[2]
+		myAddress := *e.store.GetAddress()
+
+		var left *channel.TwoPartyLedger
+		var right *channel.TwoPartyLedger
+		var ok bool
+
+		if alice != myAddress {
+			left, ok = e.store.GetTwoPartyLedger(intermediary, bob)
+			if !ok {
+				return virtualfund.VirtualFundObjective{}, fmt.Errorf("could not find a left ledger channel between %v and %v", alice, intermediary)
+			}
+		}
+		if bob != myAddress {
+			right, ok = e.store.GetTwoPartyLedger(alice, intermediary)
+			if !ok {
+				return virtualfund.VirtualFundObjective{}, fmt.Errorf("could not find a right ledger channel between %v and %v", intermediary, bob)
+			}
+		}
+
+		myRole, err := getMyRole(myAddress, participants)
+		if err != nil {
+			return virtualfund.VirtualFundObjective{}, errors.New("could not determine my role: %w")
+		}
+
+		return virtualfund.New(
+			true, // TODO ensure objective in only approved if the application has given permission somehow
+			message.SignedStates[0].State(),
+			*e.store.GetAddress(),
+			1, // Always a single hop virtual channel
+			myRole,
+			left,
+			right,
+		)
 	default:
-		return directfund.DirectFundObjective{}, errors.New("cannot handle unimplemented objective type")
+		return virtualfund.VirtualFundObjective{}, errors.New("cannot handle unimplemented objective type")
 	}
 
 }
 
 // handleLedgerRequests handles a collection of ledger requests by submitting them to the ledger manager.
-func (e *Engine) handleLedgerRequests(ledgerRequests []protocols.LedgerRequest) (protocols.SideEffects, error) {
+func (e *Engine) handleLedgerRequests(ledgerRequests []protocols.LedgerRequest, objective protocols.Objective) (protocols.SideEffects, error) {
 	sideEffects := protocols.SideEffects{}
 	for _, req := range ledgerRequests {
-		e.logger.Printf("Handling ledger request for %s", req.LedgerId)
-		ch, ok := e.store.GetChannel(req.LedgerId)
-
-		if !ok {
+		e.logger.Printf("Handling ledger request  %+v", req)
+
+		var ledger *channel.TwoPartyLedger
+		found := false
+		for _, ch := range objective.Channels() {
+			if ch.Id == req.LedgerId {
+				ledger = &channel.TwoPartyLedger{Channel: *ch}
+				found = true
+			}
+		}
+		if !found {
 			return protocols.SideEffects{}, fmt.Errorf("Could not find ledger %s", req.LedgerId)
 
 		}
-		ledger := &channel.TwoPartyLedger{Channel: *ch}
 
 		se, err := e.ledgerManager.HandleRequest(ledger, req, e.store.GetChannelSecretKey())
 		if err != nil {
 			return se, fmt.Errorf("could not handle ledger request: %w", err)
 		}
+		err = e.store.SetChannel(&ledger.Channel)
+
+		if err != nil {
+			return se, fmt.Errorf("could not set channel: %w", err)
+		}
+
 		sideEffects.Merge(se)
 	}
 	return sideEffects, nil
+}
 
+func getMyRole(myAddress types.Address, participants []types.Address) (uint, error) {
```

Comment:
add godoc

Reply:
This should almost certainly be pushed down into the protocols package. We can encapsulate **both** `n` and `myRole` entirely so that the engine doesn't need to know it https://github.com/statechannels/go-nitro/issues/257